From 1228360546a4a7e703a73aebf020f8e377d34598 Mon Sep 17 00:00:00 2001
From: Mitch Curtis <mitch.curtis@qt.io>
Date: Tue, 30 Sep 2025 11:59:42 +0800
Subject: [PATCH 104/239] SpinBox: accept touch events

If an item doesn't call setAcceptTouchEvents(true), touch events will
be converted to mouse events. This mostly worked, except for hover,
where the touch events don't translate in the way that we needed them
to.

Set this for SpinBox in the same way that
3c2eab829df900a4bcc308fe0d8d5a87cd8f56a3 did for other controls.

Fixes: QTBUG-140481
Pick-to: 6.9 6.8 6.5
Change-Id: I18751f86c1af1aea019f441a90af443cd88dea8b
Reviewed-by: Oliver Eftevaag <oliver.eftevaag@qt.io>
Reviewed-by: Shawn Rutledge <shawn.rutledge@qt.io>
(cherry picked from commit bf2ec6787c3e292c4da2c8e631248da5d9e6a357)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/quicktemplates/qquickspinbox.cpp          | 20 +++++++++----
 .../qquickmaterialstyle/data/tst_material.qml | 29 +++++++++++++++++++
 2 files changed, 43 insertions(+), 6 deletions(-)

diff --git a/src/quicktemplates/qquickspinbox.cpp b/src/quicktemplates/qquickspinbox.cpp
index 8b8f1c4d51..d1d0173820 100644
--- a/src/quicktemplates/qquickspinbox.cpp
+++ b/src/quicktemplates/qquickspinbox.cpp
@@ -343,12 +343,17 @@ bool QQuickSpinBoxPrivate::handleMove(const QPointF &point, ulong timestamp)
 {
     Q_Q(QQuickSpinBox);
     QQuickControlPrivate::handleMove(point, timestamp);
-    QQuickItem *ui = up->indicator();
-    QQuickItem *di = down->indicator();
-    up->setHovered(ui && ui->isEnabled() && ui->contains(ui->mapFromItem(q, point)));
-    up->setPressed(up->isHovered());
-    down->setHovered(di && di->isEnabled() && di->contains(di->mapFromItem(q, point)));
-    down->setPressed(down->isHovered());
+    QQuickItem *upIndicator = up->indicator();
+    const bool upIndicatorContainsPoint = upIndicator && upIndicator->isEnabled()
+        && upIndicator->contains(upIndicator->mapFromItem(q, point));
+    up->setHovered(touchId == -1 && upIndicatorContainsPoint);
+    up->setPressed(upIndicatorContainsPoint);
+
+    QQuickItem *downIndicator = down->indicator();
+    const bool downIndicatorContainsPoint = downIndicator && downIndicator->isEnabled()
+        && downIndicator->contains(downIndicator->mapFromItem(q, point));
+    down->setHovered(touchId == -1 && downIndicatorContainsPoint);
+    down->setPressed(downIndicatorContainsPoint);
 
     bool pressed = up->isPressed() || down->isPressed();
     q->setAccessibleProperty("pressed", pressed);
@@ -476,6 +481,9 @@ QQuickSpinBox::QQuickSpinBox(QQuickItem *parent)
 #if QT_CONFIG(cursor)
     setCursor(Qt::ArrowCursor);
 #endif
+#if QT_CONFIG(quicktemplates2_multitouch)
+    setAcceptTouchEvents(true);
+#endif
 }
 
 QQuickSpinBox::~QQuickSpinBox()
diff --git a/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml b/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
index 5b31f5ed84..56bf450111 100644
--- a/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
+++ b/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
@@ -1588,4 +1588,33 @@ TestCase {
         compare(popup.Material.theme, Material.Light)
         compare(popup.label.Material.theme, Material.Light)
     }
+
+    Component {
+        id: spinBoxComponent
+
+        SpinBox {
+            anchors.centerIn: parent
+        }
+    }
+
+    function test_spinBoxShouldNotBeHoveredOnTouch() {
+        let spinBox = createTemporaryObject(spinBoxComponent, testCase)
+        verify(spinBox)
+        let up = spinBox.up
+        let indicator = up.indicator
+
+        let touch = touchEvent(indicator)
+        touch.press(0, indicator, indicator.width * 0.5, indicator.height * 0.5).commit()
+        verify(up.pressed)
+        verify(!up.hovered)
+
+        // Move around a bit while pressed.
+        touch.move(0, indicator, indicator.width * 0.4, indicator.height * 0.4).commit()
+        verify(up.pressed)
+        verify(!up.hovered)
+
+        touch.release(0).commit()
+        verify(!up.pressed)
+        verify(!up.hovered)
+    }
 }
-- 
2.51.2

